### PR TITLE
Implement chapters

### LINKED ISSUE #579: Youtube Chapters support
**Is your feature request related to a usage problem (not a bug)? Please describe.**
Some videos, especially podcasts and documentaries will have chapters set up in the video that act like timestamps for the content. [Google's explanation](https://support.google.com/youtube/answer/9884579?hl=en-GB)

I'm not sure if this is technically possible but it would be nice to see as it's a great UX improvement.

**Describe the solution you'd like to see implemented**
Support for chapters.

**Describe alternatives you've considered**
n/a

**Screenshots**
[Video showing how it works](https://www.youtube.com/watch?v=b1Fo_M_tj6w&feature=emb_title), if you hover over the progress bar you can see the different chapters setup.

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by absidue**
---
Implement chapters
---

**Pull Request Type**

- [x] Feature Implementation

**Related issue**

closes #579

**Description**

This pull request implements chapters. As the Invidious API doesn't provide chapters (https://github.com/iv-org/invidious/issues/2395), the chapters are parsed from the description when the Invidious API is selected as the backend, as we don't have thumbnails the chapters will instead be shown in a compact view. This pull request also adds a distraction free setting to hide the chapters.

The chapters list can be collapsed and expanded by clicking on the header. The current chapter is shown in the header and is marked with a blue border around the thumbnail and a bold title in the list of chapters. You can also click on a chapter to seek to that timestamp.

Markers are shown on the video progress bar for the chapters.

There might still be cases where it won't parse the thumbnails for the Invidious API properly, despite me trying my best to cover every special case that I've come across, since I created the original implementation. It should just be a simple case of updating the regex to support whatever new special chapters format we've come across.

**Screenshots (if appropriate)**

<details><summary>Screenshots</summary>

chapters_collapsed_1

distraction_free

</details>

<details><summary>Local API</summary>

chapters_expanded_local_1

chapters_expanded_local_2

</details>

<details><summary>Invidious API</summary>

chapters_expanded_invidious

</details>

**Testing (for code that is not small enough to be easily understandable)**
I tested various videos with and without chapters, on both the local API and the Invidious API.

no chapters: https://youtu.be/0bvZ70wgZVM
chapters with space separator: https://youtu.be/G3EuYRbGtmA
chapters with `-` separator: https://youtu.be/0Ueel1wXnZk

**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 0.17.1

### CODE REVIEW

**Comment by PrestonN** [comment]
IIRC all videos with chapters uses information that's in the video description for determining where chapters go, so this feature should be possible to be supported with the Invidious API. There's already logic to parse the description for timestamps as well. This logic could be altered and sent back to act as chapters data.

I'd also like to see some sort of markers on the video player itself to indicate chapters. There's a few modules sitting out there that do this, but some of them might be out of date. If we need to publish a forked module in order for this to work then I'd be open to doing so.

**Comment by ChunkyProgrammer** [comment]
We do have a way of adding markers (we do this for [sponsorblock](https://github.com/FreeTubeApp/FreeTube/blob/e29d041a91fe78e96b62ac8cd61d8d8a78797ab7/src/renderer/components/ft-video-player/ft-video-player.js#L527)) but I think forking that module and allowing different types of markers so we can use it for chapters and sponsor segment markers (and make them look different) could be a good idea

**Comment by ajayyy** [comment]
There are some cases where YouTube autogenerates chapters, but most of them are pulled from the description

**Comment by absidue** [comment]
This PR is finally ready for review, I've updated the description and the screenshots with the new changes.
Major changes since you last took a look at this PR:
- Markers in the progress bar
- New compact chapters list while the Invidious API is active
- Better at extracting chapters from the description when the Invidious API is active

File: src/renderer/components/ft-video-player/ft-video-player.css [hunk 1]
```diff
@@ -4,4 +4,23 @@
...
+}
+
+:deep(.chapterMarker) {
+  height: 120%;
+  top: -10%;
+  width: 1%;
+  border-radius: 999px; /* make sure they are rounded properly*/
+  z-index: 2;
+  background-color: var(--accent-color);
 }
```
Comment (PikachuEXE) [inline src/renderer/components/ft-video-player/ft-video-player.css:26]: Why still no final newline?
We already set it in `.editorconfig`
https://github.com/FreeTubeApp/FreeTube/blob/development/.editorconfig#L9

Comment (absidue) [reply src/renderer/components/ft-video-player/ft-video-player.css]: Whoops, fixed it.

Comment (PikachuEXE) [reply src/renderer/components/ft-video-player/ft-video-player.css]: Your editor/IDE might need to be setup to support `.editorconfig`
https://editorconfig.org/#download

**Comment by PikachuEXE** [comment]
Chapters not parsed for this video:
https://youtu.be/ccg6aU7Hmjw

**Comment by absidue** [comment]
> Chapters not parsed for this video: https://youtu.be/ccg6aU7Hmjw

@PikachuEXE Can't seem to replicate it on my end, they show up for me on both API backends.

Local:
local

Invidious
invidious

**Comment by PikachuEXE** [comment]
Works for me after retrying, strange
Ignore for now

**Review by PikachuEXE — APPROVED** [review]

**Review by ChunkyProgrammer — APPROVED** [review]
LGTM

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
Me likey, LGTM! :D

**Comment by BarbzYHOOL** [comment]
This has never worked for me, is there a video showcasing it?

I'm using a remote to control freetube, so maybe it's an accessibility issue? @absidue @prestonn

**Comment by absidue** [comment]
@BarbzYHOOL Could you please explain in more detail what you mean by never worked? Do you mean that chapters aren't showing up under videos? Or is there some remote specific functionality that you were expecting.

I'm not sure what you would see in a screen recording that you can't already see in the screenshots.

**Comment by BarbzYHOOL** [comment]
@absidue 
Ho i didn't see the hidden screenshots

Ok so I never really looked below and clicked the chapters thing. Since I'm using the official youtube app for my android tv as well, I expected to see it works the same, with chapters showing ON SCREEN and not in the description, and being able to navigate them on the screen (it's very annoying to do otherwise with a remote and especially when in full screen).

So the only thing I had noticed was the small colored segments to indicate there is a chapter.

**Comment by absidue** [comment]
Always cool to hear how people are using software in ways it wasn't intended to be used. FreeTube is designed to be a desktop and mobile app, with a UI designed to be controlled with a mouse and keyboard or touch, so FreeTube not working the way TV apps work, is expected. It's like taking a racing track car onto a normal road and then complaining to the manufacturer that it can't get over speed bumps because the body is too low, it wasn't designed for that use case.

Do you have any photos of what the chapters look like in the Android TV YouTube app?

We can consider changing the chapters implementation to be more TV friendly, as long as it doesn't make the desktop or mobile experience worse, as those are our primary focus and how most people use FreeTube (this is the first time we've heard of anyone using a remote to control FreeTube).

**Comment by BarbzYHOOL** [comment]
@absidue I don't think it's very practical either with a mouse to have to scroll down and hide the video to click on the chapters.  I mean, it probably could be easier to use. I tried the current youtube desktop method, it appears on the right, which is easier to use. If you want to know what are the other awesome features of their official app (which has really improved in like 3 years) I can tell you some that I find nice.

For screenshots, it's hard to find. here is what I found:
https://images.frandroid.com/wp-content/uploads/2021/12/chapitres-youtube-frandroid-1-1200x737.jpg
https://tekdeeps.com/youtube-will-also-display-video-chapters-in-the-smart-tv-and-game-console-app/

I think it has been updated a little since then. So what it does is that you press the "up" arrow, it displays the timeline of the video with the chapters sections and current chapter like in the screenshots, then you press the right/left arrow and it pauses the video and lets you navigate the timeline, if you press the up arrow again you can navigate through the thumbnails of each chapter (i found it alone, so it means it's not even hard to use) and it jumps on the timeline as well, press OK and it jumps to it. You don't have to leave the video at all, it would work in full screen on a desktop.
I use it all the time.

Btw, freetube is almost perfect with a remote, it doesn't require much more. Maybe more people than me are using it like that, I use it for 1 year already and never posted before!  Here is the remote I use which is not the best (but hard to know that before you have tried one for the first time, lacking that scroll wheel button so much!!) https://ae01.alicdn.com/kf/HTB1CPguhhuTBuNkHFNRq6A9qpXaY/Rii-Mini-clavier-sans-fil-2-4GHz-i24-pour-Android-TV-Box-Mini-PC-ordinateur-portable.jpg_Q90.jpg_.webp

**Pull request merged**